include/os/linux/spl/sys/shrinker.h

```diff
@@ -32,46 +32,35 @@
  * Due to frequent changes in the shrinker API the following
  * compatibility wrappers should be used.  They are as follows:
  *
- * SPL_SHRINKER_DECLARE is used to declare the shrinker which is
- * passed to spl_register_shrinker()/spl_unregister_shrinker().  Use
- * shrinker_name to set the shrinker variable name, shrinker_callback
- * to set the callback function, and seek_cost to define the cost of
- * reclaiming an object.
+ *   SPL_SHRINKER_DECLARE(varname, countfunc, scanfunc, seek_cost);
  *
- *   SPL_SHRINKER_DECLARE(shrinker_name, shrinker_callback, seek_cost);
- *
- * SPL_SHRINKER_CALLBACK_FWD_DECLARE is used when a forward declaration
- * of the shrinker callback function is required.  Only the callback
- * function needs to be passed.
- *
- *   SPL_SHRINKER_CALLBACK_FWD_DECLARE(shrinker_callback);
- *
- * SPL_SHRINKER_CALLBACK_WRAPPER is used to declare the callback function
- * which is registered with the shrinker.  This function will call your
- * custom shrinker which must use the following prototype.  Notice the
- * leading __'s, these must be appended to the callback_function name.
- *
- *   int  __shrinker_callback(struct shrinker *, struct shrink_control *)
- *   SPL_SHRINKER_CALLBACK_WRAPPER(shrinker_callback);a
+ * SPL_SHRINKER_DECLARE is used to declare a shrinker with the name varname,
+ * which is passed to spl_register_shrinker()/spl_unregister_shrinker().
+ * The countfunc returns the number of free-able objects.
+ * The scanfunc returns the number of objects that were freed.
  *
  *
  * Example:
  *
- * SPL_SHRINKER_CALLBACK_FWD_DECLARE(my_shrinker_fn);
- * SPL_SHRINKER_DECLARE(my_shrinker, my_shrinker_fn, 1);
- *
- * static int
- * __my_shrinker_fn(struct shrinker *shrink, struct shrink_control *sc)
+ * static unsigned long
+ * my_count(struct shrinker *shrink, struct shrink_control *sc)
  * {
- *	if (sc->nr_to_scan) {
- *		...scan objects in the cache and reclaim them...
- *	}
- *
  *	...calculate number of objects in the cache...
```

Comment:
I think that this commit maintains all of the existing behavior, however incorrect.  Let me know if you see anything that contradicts that.

The callbacks return SHRINK_STOP if they don't want to be called again.  For the older shared interface, this is `#define`'d to `-1`.

Are the 3 interfaces you're referring to:
1. `<=3.10`, single callback
2. `>3.12, <4.19`, two callbacks, can return SHRINK_STOP
3. `>4.19`, two callbacks, can return SHRINK_STOP or SHRINK_EMPTY

So if we wanted to return SHRINK_EMPTY, we would need to think about what it would do in cases 1 and 2 above.  We don't do that today, but I agree we would need more infrastructure to do so cleanly.  Do you think we should add a comment explaining that the callbacks can return SHRINK_STOP but not SHRINK_EMPTY?

Reply:
I agree, it does maintain the existing behavior.  Upon closer inspection I believe that behavior is also correct for both interfaces.  Regarding SHRINK_EMPTY, I'd definitely be in favor of adding a comment that it's not currently supported.

Reply:
Sounds good, updated.  FYI, if a callback tried to use SHRINK_EMPTY, it wouldn't compile on the older kernels, so buildbot would catch it.